### PR TITLE
[UDOP] Add special tokens to tokenizer

### LINKED ISSUE #29591: Special tokens of UDOP aren't encoded/decoded correctly
### System Info

Transformers dev 4.38.2

### Who can help?

Question for @ArthurZucker

### Reproduction

UDOP has been added and the model works as expected. However for the tokenizer, the authors defined [1201 special tokens](https://github.com/microsoft/i-Code/blob/4f268cbe88bd6f0e99840f766cf4429feaa8aee8/i-Code-Doc/core/models/tokenization.py#L17-L43) (only used during pre-training) which need to be added manually to the vocabulary of the tokenizer in order to get one-on-one matching results with the original implementation.

Based on comment here: https://github.com/huggingface/transformers/pull/22940#discussion_r1433053841, I tried doing this:

```python
from transformers import UdopTokenizer

# Add extra_ids to the special token list
extra_ids = 100
loc_extra_ids = 501
other_extra_ids = 200

additional_special_tokens = []
if extra_ids > 0 and "<extra_id_0>" not in additional_special_tokens:
    additional_special_tokens = ["<extra_id_{}>".format(i) for i in range(extra_ids)]
    additional_special_tokens.extend(["<extra_l_id_{}>".format(i) for i in range(extra_ids)])
    additional_special_tokens.extend(["</extra_l_id_{}>".format(i) for i in range(extra_ids)])
    additional_special_tokens.extend(["<extra_t_id_{}>".format(i) for i in range(extra_ids)])
    additional_special_tokens.extend(["</extra_t_id_{}>".format(i) for i in range(extra_ids)])

if loc_extra_ids > 0 and "<loc_0>" not in additional_special_tokens:
    additional_special_tokens.extend(["<loc_{}>".format(i) for i in range(loc_extra_ids)])

if other_extra_ids > 0 and "<other_0>" not in additional_special_tokens:
    additional_special_tokens.extend(["<other_{}>".format(i) for i in range(other_extra_ids)])

hf_tokenizer = UdopTokenizer.from_pretrained("t5-base", legacy=True, additional_special_tokens=additional_special_tokens)
```

As this could avoid the [regex](https://github.com/microsoft/i-Code/blob/4f268cbe88bd6f0e99840f766cf4429feaa8aee8/i-Code-Doc/core/models/tokenization.py#L67-L98) defined by the authors when encoding. However, this doesn't result in the same behaviour. Let's say you have the tensor [0,  8986, 32942, 32966, 32554, 32551, 1]. The original tokenizer decodes this to (using the [original notebook](https://github.com/microsoft/i-Code/blob/main/i-Code-Doc/example_io.ipynb)):

```
from core.models import UdopTokenizer

original_tokenizer = UdopTokenizer.from_pretrained("path_to_original_files")

print(original_tokenizer.decode([0,  8986, 32942, 32966, 32554, 32551, 1]))
>>> paragraph<loc_58><loc_34><loc_446><loc_449>
```
whereas with the tokenizer defined above we get:

```
print(hf_tokenizer.decode([0,  8986, 32942, 32966, 32554, 32551, 1]))
>>> paragraph <loc_442> <loc_466> <loc_54> <loc_51>
```
This is due to [this function](https://github.com/microsoft/i-Code/blob/4f268cbe88bd6f0e99840f766cf4429feaa8aee8/i-Code-Doc/core/models/tokenization.py#L100), I assume. This causes the additional tokens to not have integers that are incremental:

```
for token in original_tokenizer.additional_special_tokens:
  print(token, original_tokenizer.convert_tokens_to_ids([token]))
```

```
<extra_id_0> [32099]
<extra_id_1> [32098]
<extra_id_2> [32097]
<extra_id_3> [32096]
<extra_id_4> [32095]
<extra_id_5> [32094]
<extra_id_6> [32093]
<extra_id_7> [32092]
<extra_id_8> [32091]
<extra_id_9> [32090]
<extra_id_10> [32089]
<extra_id_11> [32088]
<extra_id_12> [32087]
<extra_id_13> [32086]
```

### Expected behavior

Equivalent encoding/decoding of the 1201 special tokens with the original implementation

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR makes sure the 1201 additional special tokens can be encoded/decoded properly.

Fixes #29591

### CODE REVIEW

File: tests/models/udop/test_tokenization_udop.py [hunk 1]
```diff
@@ -1888,3 +1888,13 @@ def test_text_target(self):
...
         self.assertDictEqual(dict(encoding_p), dict(encoding_r))
         self.assertEqual(tokenizer_p.decode(encoding_p["input_ids"]), expected_decoding)
+
+    def test_special_tokens(self):
+        tokenizer_r = UdopTokenizerFast.from_pretrained("microsoft/udop-large")
+
+        ids = [0, 8986, 32942, 32966, 32554, 32551, 1]
+        decoding = tokenizer_r.decode(ids)
+
+        excepted_decoding = "<pad> paragraph<loc_58><loc_34><loc_446><loc_449></s>"
```
Comment (NielsRogge) [inline tests/models/udop/test_tokenization_udop.py:1898]: @ArthurZucker I've noticed that the slow Python tokenizer returns this:

```
<pad>paragraph <loc_58> <loc_34> <loc_446> <loc_449> </s>
```
in other words, with spaces in between the special tokens. Does this have anything to do with rstrip/lstrip?

Comment (ArthurZucker) [reply tests/models/udop/test_tokenization_udop.py:1898]: `spaces_between_special_tokens=True` is the defaults 😢

File: src/transformers/models/udop/convert_udop_to_hf.py [hunk 2]
```diff
@@ -119,13 +119,29 @@ def convert_udop_checkpoint(model_name, pytorch_dump_folder_path=None, push_to_h
...
+    # fmt: on
+
+    # additional_special_tokens = [
+    #     AddedToken(token, rstrip=False, lstrip=False, normalized=False, special=True)
+    #     for token in additional_special_tokens
+    # ]
```
Comment (ArthurZucker) [inline src/transformers/models/udop/convert_udop_to_hf.py:131]:
```suggestion
```

File: tests/models/udop/test_tokenization_udop.py (hunk 1 above)
Comment (ArthurZucker) [reply tests/models/udop/test_tokenization_udop.py:1898]: make sure to test both!

**Review by ArthurZucker — COMMENTED** [review]
LGTM, let's do better testing: 
`paragraph<loc_58>. Hey` for example with a special case

**Comment by NielsRogge** [comment]
Thanks, I've added tests. Should we add `spaces_between_special_tokens` as argument to `UdopTokenizer`, to make it decode the same as the fast tokenizer?

File: tests/models/udop/test_tokenization_udop.py (hunk 1 above)
Comment (ArthurZucker) [reply tests/models/udop/test_tokenization_udop.py:1917]: We should test `tokenizer_p.decode(ids, spaces_between_special_tokens = False)`. If that is not the issue it's decoding function that is at fault!

Comment (NielsRogge) [reply tests/models/udop/test_tokenization_udop.py:1917]: Ok I've tested this, but it results in different decodings:
```
assert decoding_p == decoding_r == excepted_decoding
        AssertionError: assert '<pad>paragra...<loc_449></s>' == '<pad> paragr...<loc_449></s>'
          - <pad> paragraph<loc_58><loc_34><loc_446><loc_449></s>
          ?      -
         + <pad>paragraph<loc_58><loc_34><loc_446><loc_449></s>
```
Looks like the slow tokenizer adds a space after the /<pad> token

**Comment by ArthurZucker** [comment]
You can add it and default it to `False` in the `decode` maybe? Cuz I don't think `decode` takes `self. spaces_between_special_tokens` into account no?

**Comment by NielsRogge** [comment]
Yeah indeed, shouldn't the `_decode` method [here](https://github.com/huggingface/transformers/blob/caa5c65db1f4db617cdac2ad667ba62edf94dd98/src/transformers/tokenization_utils.py#L1026) take into account `self. spaces_between_special_tokens` in case it exists?

**Review by ArthurZucker — APPROVED** [review]
LGTM, breaking but fixing

**Comment by NielsRogge** [comment]
I don't think it's a breaking change, since adding spaces_between_special_tokens as an attribute doesn't do anything. The _decode [method](https://github.com/huggingface/transformers/blob/e4ea19b958c89d61e42461fac6ac8441787121f8/src/transformers/tokenization_utils.py#L991-L997) doesn't use it (whereas I think it should).

**Comment by ArthurZucker** [comment]
If it does not do anything why are we adding it?

**Comment by NielsRogge** [comment]
Haha yes I'll remove it, and I'll remove it for Gemma too

**Pull request merged**